File: mobile/lib/data/models/purchase_order.dart
class PurchaseOrder {
  final String id;
  final String poNumber;
  final String status;
  final int totalCents;
  final String currency;
  final String? vendorId;
  final String? vendorName;
  final String? issuedAt;
  final String? expectedDeliveryDate;
  final String? createdAt;
  final List<POLineItem> lineItems;

  const PurchaseOrder({
    required this.id,
    required this.poNumber,
    required this.status,
    required this.totalCents,
    this.currency = 'INR',
    this.vendorId,
    this.vendorName,
    this.issuedAt,
    this.expectedDeliveryDate,
    this.createdAt,
    this.lineItems = const [],
  });

  factory PurchaseOrder.fromJson(Map<String, dynamic> json) {
    return PurchaseOrder(
      id: json['id'] ?? '',
      poNumber: json['po_number'] ?? json['poNumber'] ?? '',
      status: json['status'] ?? '',
      totalCents: json['total_cents'] ?? json['totalCents'] ?? 0,
      currency: json['currency'] ?? 'INR',
      vendorId: json['vendor_id'],
      vendorName: json['vendor_name'],
      issuedAt: json['issued_at'],
      expectedDeliveryDate: json['expected_delivery_date'],
      createdAt: json['created_at'],
      lineItems:
          (json['line_items'] as List<dynamic>?)
              ?.map((e) => POLineItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}

class POLineItem {
  final String id;
  final String description;
  final double quantity;
  final int unitPriceCents;

  const POLineItem({
    required this.id,
    required this.description,
    required this.quantity,
    required this.unitPriceCents,
  });

  factory POLineItem.fromJson(Map<String, dynamic> json) {
    return POLineItem(
      id: json['id'] ?? '',
      description: json['description'] ?? '',
      quantity: (json['quantity'] ?? 0).toDouble(),
      unitPriceCents: json['unit_price_cents'] ?? json['unitPriceCents'] ?? 0,
    );
  }
}

File: api/models/purchase_order.py
import uuid
from datetime import datetime, date
from typing import Optional

from sqlalchemy import (
    String,
    BigInteger,
    Integer,
    DateTime,
    Date,
    Text,
    ForeignKey,
    UniqueConstraint,
    CheckConstraint,
    Index,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from api.database import Base


class PurchaseOrder(Base):
    __tablename__ = "purchase_orders"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False
    )
    po_number: Mapped[str] = mapped_column(
        String(50), unique=True, nullable=False
    )
    pr_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("purchase_requests.id")
    )
    vendor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=False
    )
    status: Mapped[str] = mapped_column(String(50), default="DRAFT")
    total_cents: Mapped[int] = mapped_column(BigInteger, nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="INR")
    expected_delivery_date: Mapped[Optional[date]] = mapped_column(Date)
    terms_and_conditions: Mapped[Optional[str]] = mapped_column(Text)
    issued_at: Mapped[Optional[datetime]] = mapped_column(DateTime)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )
    deleted_at: Mapped[Optional[datetime]] = mapped_column(DateTime)

    __table_args__ = (
        Index("idx_po_tenant", "tenant_id"),
        Index("idx_po_vendor", "vendor_id"),
        Index("idx_po_status", "status"),
    )


class PoLineItem(Base):
    __tablename__ = "po_line_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    po_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("purchase_orders.id", ondelete="CASCADE"),
        nullable=False,
    )
    line_number: Mapped[int] = mapped_column(Integer, nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    unit_price_cents: Mapped[int] = mapped_column(BigInteger, nullable=False)
    received_quantity: Mapped[int] = mapped_column(Integer, default=0)

    __table_args__ = (
        UniqueConstraint("po_id", "line_number", name="uq_po_line_item"),
        CheckConstraint("quantity > 0", name="chk_po_line_qty"),
        CheckConstraint(
            "unit_price_cents > 0", name="chk_po_line_price"
        ),
        Index("idx_po_items_po", "po_id"),
    )
